dnf5/commands/provides/provides.cpp

```diff
@@ -0,0 +1,145 @@
+                pkg_specs.emplace_back(argv[i]);
+            }
+            return true;
+        });
+    keys->set_complete_hook_func([&ctx](const char * arg) { return match_specs(ctx, arg, false, true, true, false); });
+    cmd.register_positional_arg(keys);
+}
+
+void ProvidesCommand::configure() {
+    auto & context = get_context();
+    context.set_load_system_repo(true);
+    context.base.get_config().get_optional_metadata_types_option().add_item(
+        libdnf5::Option::Priority::RUNTIME, libdnf5::METADATA_TYPE_FILELISTS);
+    context.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+}
+
+void ProvidesCommand::run() {
+    auto & ctx = get_context();
+    bool any_match = false;
+    std::set<std::string> unmatched_specs;
+
+    for (auto & spec : pkg_specs) {
+        libdnf5::rpm::PackageQuery full_package_query(ctx.base);
+        auto provides_query = full_package_query;
+        auto filename_query = full_package_query;
+        auto binary_query = full_package_query;
+        int match = 0;
+
+        do {
+            // if the spec starts with "/" assume it's a provide by filename and skip this query
+            if (!(spec.rfind("/", 0) == 0)) {
+                provides_query.filter_provides(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+                if (!provides_query.empty()) {
+                    full_package_query = provides_query;
+                    match = libdnf5::cli::output::ProvidesMatchedBy::PROVIDES;
+                    break;
+                }
+            }
+            // compatibility for packages that didn't do UsrMove
+            if ((spec.rfind("/bin/", 0) == 0) || (spec.rfind("/sbin/", 0) == 0)) {
+                spec.insert(0, "/usr");
+            }
+            filename_query.filter_file(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+            if (!filename_query.empty()) {
+                full_package_query = filename_query;
+                match = libdnf5::cli::output::ProvidesMatchedBy::FILENAME;
+                break;
+            }
+            // if the spec is a binary name and there is no match from provide, try to prepend
+            // /bin /sbin to match it by binary provides and continue
+            else {
+                std::string bin_spec = "/bin/" + spec;
+                std::string sbin_spec = "/sbin/" + spec;
+                std::string usr_bin_spec = "/usr/bin/" + spec;
+                std::string usr_sbin_spec = "/usr/sbin/" + spec;
+                binary_query.filter_file(
+                    std::vector<std::string>({bin_spec, sbin_spec, usr_bin_spec, usr_sbin_spec}),
+                    libdnf5::sack::QueryCmp::GLOB);
+                if (!binary_query.empty()) {
+                    full_package_query = binary_query;
+                    match = libdnf5::cli::output::ProvidesMatchedBy::BINARY;
+                    break;
+                }
+            }
+        } while (match);  // do
```

Comment:
Hmm, I find the `do {} while;` approach confusing because it's not a typical loop and involves many jumps (breaks), making the workflow hard to follow. Though I eventually understand it, it's a bit painful 😃

There are three blocks representing different match types—`filter_by_provides`, `filter_by_filename`, and `filter_by_binary`. Consider combining them into a single method, like `filter_spec`, which internally handles each condition. This way, the `run()` function can simply call `filter_spec`, making the code more straightforward. What do you think? 🙂

Reply:
I thought about the chunkiness too, but I wanted to avoid `goto`s and similar stuff. I   chose do-while because it's kinda better imho to skip to some execution. My idea for the implementation was, you match -> you break.

One problem I had to face while writing this code is that the matching becomes gradually more expensive. so match-by-provides takes less time than match-by-filename which again takes less than match-by-binary. so, yes, it's a bit convoluted but if we want to have the same behavior like dnf4 and try to do things smart I think we have to cut the niceness somewhere. still not sure if there are better ways to achieve the desired results.

I like the `filter_spec` approach, though, to make `run` much more readable. I'll do it.

Reply:
strongly agree

Reply:
But still, if you're not convinced about this approach, feel free to keep it as it is. I might be biased due to my experience with TDD :P